### PR TITLE
Refactor Sigdata methods with sigset

### PR DESCRIPTION
**Pull request by tnachen**
Changed stack_ret_arg to become a boolean on SigData, as the stack_return_arg is always the last argument.
Refactored all SigSet methods on SigData to be in SigSet instead.

### CODE REVIEW

**Comment by tnachen** [comment]
@jameysharp changing the Option<u16> to bool made the SigData size from 40 -> 32.
However, I'm trying to figure out how to remove more fields based on the SigSet suggestion. 
Based on your earlier comment, seems like the motivation is to be able to index over SigDatas since they're all lined up in the abi args list.
But from the function calls it seems like we're parsing a function sig one at a time and not sure what the change should look like here.

**Review by jameysharp — COMMENTED** [review]
I'll defer to @fitzgen and @cfallin on whether it's okay to make `stack_ret_arg` a `bool`. You've shown that every ABI we currently support places that argument at the end, which I think is a pretty good argument for doing this. But we want to support more targets in the future, so maybe we don't want to over-simplify this detail? On the other hand, we can always change this again if we ever find a target that doesn't put the stack ret arg at the end.

I think we should at least merge your changes that replace, for example, `self.lower_ctx.sigs()[*abi].get_ret_arg(self.lower_ctx.sigs())` with `self.lower_ctx.sigs().get_ret_arg(abi)`. That's so much nicer, even if we don't shrink `SigData` any further!

The next step from there is to remove from `SigData` the `args_length`/`rets_length` fields… or remove the `args_start`/`rets_start` fields and turn the length fields back into `args_end`/`rets_end`, which I think will be better. When constructing a new `SigData`, just delete anything related to setting the removed fields.

When you need to access the removed fields, you'll get them instead from the neighboring `SigData`. If you have `args_end` and you need `args_start`, then you need to access `sig-1` and get its `args_end`—except for the very first `sig`, where `args_start` is always 0.

Now, I said `sig-1`, but `Sig` is a struct wrapper around a `u32`, so you can't just do arithmetic on it. I'd add a little helper function for this, something like:

```rust
impl Sig {
    fn prev(self) -> Option<Sig> {
        self.index().checked_sub(1).map(Sig::new)
    }
}
```

Then something like this should work:

```rust
let args_start = sig.prev().map_or(0, |prev| self.sigs[prev].args_end);
```

Does that answer your question?

**Comment by bjorn3** [comment]
> I'll defer to @fitzgen and @cfallin on whether it's okay to make stack_ret_arg a bool. You've shown that every ABI we currently support places that argument at the end, which I think is a pretty good argument for doing this. But we want to support more targets in the future, so maybe we don't want to over-simplify this detail? On the other hand, we can always change this again if we ever find a target that doesn't put the stack ret arg at the end.

The position shouldn't matter I think. Only the register assignment.

**Comment by jameysharp** [comment]
I talked with @fitzgen and @cfallin about this offline. We aren't comfortable hard-coding the assumption that `stack_return_arg` is always last right now. (@bjorn3, you're right that the register assignment is all that matters, but as I understand it the register assignment gets computed from this list, so order still matters.)

So @tnachen, could you remove the `stack_return_arg` changes from this PR? I want to merge the rest of the work you did here. Let me know whether you want to try de-duplicating the `args_start` fields in this PR, or if I should merge this and you can open another PR. I would recommend opening another PR for that unless you feel you can do it within a day or two, just so we can get the work you've already done merged promptly.

**Comment by tnachen** [comment]
@jameysharp ah ok! I just reverted those changes and kept the Sigdata refactoring. 
We can merge this as-is and I can submit another one for the return field changes

File: cranelift/codegen/src/machinst/abi.rs [hunk 1]
```diff
@@ -1619,7 +1621,8 @@ impl<M: ABIMachineSpec> Callee<M> {
         sigs: &SigSet,
         vregs: &mut VRegAllocator<M::I>,
     ) -> Option<M::I> {
-        if let Some(i) = sigs[self.sig].stack_ret_arg {
+        let sig_data = &sigs[self.sig];
+        if let Some(i) = sig_data.stack_ret_arg {
```
Comment (jameysharp) [inline cranelift/codegen/src/machinst/abi.rs:1625]: This is a small thing, but could you put this back the way it was before?

```suggestion
        if let Some(i) = sigs[self.sig].stack_ret_arg {
```

File: cranelift/codegen/src/machinst/abi.rs [hunk 2]
```diff
@@ -920,6 +797,137 @@ impl SigSet {
...
+            rets_length,
+            sized_stack_arg_space,
+            sized_stack_ret_space,
+            has_stack_ret_arg: need_stack_return_area,
+            call_conv: sig.call_conv,
+        })
+    }
+
+    /// Get this signature's ABI arguments.
+    pub fn args(&self, sig: &Sig) -> &[ABIArg] {
```
[the remark this answers is not in the document]
Comment (jameysharp) [reply cranelift/codegen/src/machinst/abi.rs:861]: I would like you to remove the `&` from all the `&Sig` parameters, which will mean removing some `*` dereferences in a few places too. I hope that's easy to do, but let me know if it turns out to be complicated for some reason I haven't thought of.

File: cranelift/codegen/src/machinst/abi.rs [hunk 3]
```diff
@@ -830,7 +705,6 @@ pub struct SigSet {
     /// The actual ABI signatures, keyed by `Sig`.
     sigs: PrimaryMap<Sig, SigData>,
 }
-
```
Comment (jameysharp) [inline cranelift/codegen/src/machinst/abi.rs:833]: I wouldn't bother to mention this except I'm making some other requests already: could you put this blank line back?

**Review by jameysharp — COMMENTED** [review]
This looks great! I just have a few small details I'd appreciate if you could clean up.

**Comment by tnachen** [comment]
@jameysharp addressed comments!

**Review by jameysharp — APPROVED** [review]
Thank you for your patience with my code-review requests!

**Pull request merged**